crates/tokio-util/src/event_listeners.rs

```diff
@@ -1,46 +1,59 @@
-use tokio::sync::mpsc;
-use tokio_stream::wrappers::UnboundedReceiverStream;
+use tokio::sync::broadcast::{self, Sender};
+use tokio_stream::wrappers::BroadcastStream;
+use tracing::{error, warn};
 
-/// A collection of event listeners for a task.
-#[derive(Clone, Debug)]
+const DEFAULT_BROADCAST_CHANNEL_SIZE: usize = 1000;
+
+/// A bounded broadcast channel for a task.
+#[derive(Debug)]
 pub struct EventListeners<T> {
-    /// All listeners for events
-    listeners: Vec<mpsc::UnboundedSender<T>>,
+    /// The sender part of the broadcast channel
+    sender: Sender<T>,
+}
+
+impl<T: Clone> Clone for EventListeners<T> {
+    fn clone(&self) -> Self {
+        EventListeners { sender: self.sender.clone() }
+    }
 }
 
-impl<T> Default for EventListeners<T> {
+impl<T: Clone + Send + Sync + 'static> Default for EventListeners<T> {
     fn default() -> Self {
-        Self { listeners: Vec::new() }
+        Self::new(DEFAULT_BROADCAST_CHANNEL_SIZE)
     }
 }
 
-impl<T: Clone> EventListeners<T> {
-    /// Send an event to all listeners.
-    ///
-    /// Channels that were closed are removed.
-    pub fn notify(&mut self, event: T) {
-        self.listeners.retain(|listener| listener.send(event.clone()).is_ok())
+impl<T: Clone + Send + Sync + 'static> EventListeners<T> {
+    /// Creates a new `EventListeners`.
+    pub fn new(broadcast_channel_size: usize) -> Self {
+        let (sender, _) = broadcast::channel(broadcast_channel_size);
+        Self { sender }
```

Comment:
dropping the first receiver is a bug, see docs
> If all Receiver handles are dropped, the send method will return a SendError.

https://docs.rs/tokio/latest/tokio/sync/broadcast/fn.channel.html

this method needs to return the receiver, and can hence only be called when the receiver is needed and can stored somewhere

Reply:
yep, it is made this way on purpose, it mimics the previous behaviour we had, in which after creating an `EventListeners` instance, the vector of listeners is empty and we can call notify. The consumers of EventListeners are supposed to add listeners before trying to notify them (this is what they do now, nothing has changed on the consumer side).

In case we try to send and there are no subscribers, the SendError returned would be handled and logged in the `notify` method, let me know wdyt

Reply:
ah, it's recoverable https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c1f6a86a97d54e601509708dd3413051